cluster.go

```diff
@@ -570,6 +573,23 @@ func (c *ClusterClient) cmdSlotAndNode(state *clusterState, cmd Cmder) (int, *cl
 	return slot, node, err
 }
 
+func (c *ClusterClient) Context() context.Context {
+	if c.ctx != nil {
+		return c.ctx
+	}
+	return context.Background()
+}
+
+// WithContext changes the ClusterClient's context to ctx and returns
+// a reference to the ClusterClient.
+func (c *ClusterClient) WithContext(ctx context.Context) *ClusterClient {
+	if ctx == nil {
+		panic("nil context")
+	}
+	c.ctx = ctx
```

Comment:
Please note that this method doesn't return a copy anymore, it now returns the same instance. There is a problem with copying the ClusterClient and Ring because they contain `sync.Mutex`. I'm tempted to suggest to remove the copying from the regular `Client` too, but I'm afraid of how that would potentially affect users.

Reply:
It's true that some libraries return a copy (e.g. [golang.org/x/sync/errgroup](https://github.com/golang/dep/blob/73e9c7562b3447bc130f9fe7b2bbe6eeae6d173c/vendor/golang.org/x/sync/errgroup/errgroup.go#L33-L36)), while others return a reference to the same object (e.g. [github.com/gocql/gocql](https://github.com/gocql/gocql/blob/697e7c57f99bb127606c4d4f975c2ff4f34eca1c/session.go#L767-L770)). So I guess we could potentially go for both options.